auth/user.go

```diff
@@ -191,32 +189,66 @@ func (user *userImpl) SetExplicitRoles(roles ch.TimedSet) {
 func (user *userImpl) Authenticate(password string) bool {
 	if user == nil {
 		return false
-	} else if user.OldPasswordHash_ != nil {
+	}
+
+	// exit early for disabled user accounts
+	if user.Disabled_ {
+		return false
+	}
+
+	// exit early if old hash is present
+	if user.OldPasswordHash_ != nil {
 		base.Warnf(base.KeyAll, "User account %q still has pre-beta password hash; need to reset password", base.UD(user.Name_))
 		return false // Password must be reset to use new (bcrypt) password hash
-	} else if user.PasswordHash_ == nil {
+	}
+
+	// bcrypt hash present
+	if user.PasswordHash_ != nil {
+		if !compareHashAndPassword(user.PasswordHash_, []byte(password)) {
+			// incorrect password
+			return false
+		}
+
+		// password was correct, we'll attempt to upgrade the bcrypt hash
+		user.upgradePasswordHash(user.PasswordHash_, password)
+	} else {
+		// no hash, but (incorrect) password provided
 		if password != "" {
 			return false
 		}
-	} else if !compareHashAndPassword(user.PasswordHash_, []byte(password)) {
-		return false
 	}
-	return !user.Disabled_
+
+	return true
 }
 
 // Changes a user's password to the given string.
 func (user *userImpl) SetPassword(password string) {
 	if password == "" {
 		user.PasswordHash_ = nil
 	} else {
-		hash, err := bcrypt.GenerateFromPassword([]byte(password), kBcryptCostFactor)
+		hash, err := bcrypt.GenerateFromPassword([]byte(password), bcryptCost)
 		if err != nil {
 			panic(fmt.Sprintf("Error hashing password: %v", err))
 		}
 		user.PasswordHash_ = hash
 	}
 }
 
+// upgraePasswordHash will check the bcrypt cost of the given hash
+// and will reset the user's password if the configured cost has since increased
+// Callers should make sure password is correct before calling this
+func (user *userImpl) upgradePasswordHash(hash []byte, password string) {
+	hashCost, costErr := bcrypt.Cost(user.PasswordHash_)
+	if costErr == nil && hashCost < bcryptCost {
```

Comment:
Do we need to consider users tuning this *down*, for example testing and deciding they want to drop it back to the default?

Reply:
I guess it's feasible that a customer sets bcrypt.MaxCost for "better security" and now it takes x hours to hash a single uncached password...

I'm wondering if we limit the ratchet up to the default cost, and then allow upgrades and downgrades above that?

On a typical 2.2GHz i7 these are the times for each bcrypt cost value:
```
BenchmarkBcryptCostTime/cost4-8                      1415281 ns/op (1.4 ms) <-- bcrypt.MinCost
BenchmarkBcryptCostTime/cost5-8                      2608134 ns/op
BenchmarkBcryptCostTime/cost6-8                      4861873 ns/op
BenchmarkBcryptCostTime/cost7-8                      9799124 ns/op
BenchmarkBcryptCostTime/cost8-8                     19212657 ns/op
BenchmarkBcryptCostTime/cost9-8                     40039870 ns/op
BenchmarkBcryptCostTime/cost10-8                    76995010 ns/op (76 ms) <-- bcrypt.DefaultCost
BenchmarkBcryptCostTime/cost11-8                   167217623 ns/op
BenchmarkBcryptCostTime/cost12-8                   314051218 ns/op
BenchmarkBcryptCostTime/cost13-8                   622015597 ns/op
BenchmarkBcryptCostTime/cost14-8                  1241824291 ns/op 
BenchmarkBcryptCostTime/cost15-8                  2495447742 ns/op
BenchmarkBcryptCostTime/cost16-8                  4997318661 ns/op
BenchmarkBcryptCostTime/cost17-8                  9956218788 ns/op
BenchmarkBcryptCostTime/cost18-8                 20137031016 ns/op
BenchmarkBcryptCostTime/cost19-8                 40076835984 ns/op
BenchmarkBcryptCostTime/cost20-8                 80906926691 ns/op
BenchmarkBcryptCostTime/cost21-8                161161571484 ns/op (2.5 min) <-- limit of my patience
^Csignal: interrupt
```

Reply:
Actually I'm erring on the side of caution, and think we shouldn't allow the user to configure a bcrypt_cost lower than default at all.

Reply:
Yeah, I think setting the current default as the lower bound would be good :).